python/cudf/cudf/core/mixins/mixin_factory.py

```diff
@@ -0,0 +1,169 @@
+    ... )
+
+    >>> # The above is equivalent to defining a class like so:
+    ... class MyFoo:
+    ...     # The set of valid foo operations.
+    ...     _VALID_FOOS = {"foo1", "foo2"}
+    ...
+    ...     # This is the method for child classes to override. Note that the
+    ...     # first parameter is always called "op".
+    ...     def do_foo(self, op, *args, **kwargs):
+    ...         raise NotImplementedError
+
+    >>> # MyFoo can be used as follows.
+    >>> class BarImplementsFoo(MyFoo):
+    ...     _SUPPORTED_FOOS = ("foo1",)
+    ...     _FOO_DOCSTRINGS = {"ret": "42"}
+    ...
+    ...     # This method's docstring will be formatted and used for all valid
+    ...     # operations. The `op` key is always automatically filled in, while
+    ...     # other keys are formatted using _FOO_DOCSTRINGS.
+    ...     def do_foo(self, op, *args, **kwargs):
+    ...         '''Perform the operation {op}, which returns {ret}.'''
+    ...         return 42
+
+    >>> bar = BarImplementsFoo()
+    >>> print(bar.foo1())
+    42
+
+    >>> # This will raise an AttributeError because foo2 is not supported by
+    >>> # the BarImplementsFoo subclass of MyFoo.
+    >>> # print(bar.foo2())
+
+    >>> # The docstring is formatted with the operation name as well as any
+    >>> # additional keys provided via the _FOO_DOCSTRINGS parameter.
+    >>> print(bar.foo1.__doc__)
+    Perform the operation foo1, which returns 42.
+    """
+
+    docstring_attr = f"{category_name}_DOCSTRINGS"
+    validity_attr = f"_VALID_{category_name}S"
+
+    class OperationMixin:
+        @classmethod
+        def _add_operation(cls, operation):
+            # This function creates operations on-the-fly.
+
+            # Generate a signature without the `op` parameter.
+            base_operation = getattr(cls, operation_name)
+            base_params = inspect.signature(base_operation).parameters.values()
+            params = [p for p in base_params if p.name != "op"]
+            signature = inspect.Signature(parameters=params)
+
+            # Generate the list of arguments forwarded to the base operation.
+            arglist = ", ".join(
+                [
+                    f"{key}={key}"
+                    for key in signature.parameters
+                    if key not in ("self", "args", "kwargs")
+                ]
+            )
+            arglist = arglist + (", " if arglist else "") + "*args, **kwargs"
+
+            # Apply the formatted docstring of the base operation to the
+            # operation being created here.
+            docstring = base_operation.__doc__.format(
+                cls=cls.__name__,
+                op=operation,
+                **getattr(cls, docstring_attr, {}).get(operation, {}),
+            )
+
+            namespace = {}
+            exec_str = f"""
+def {operation}{str(signature)}:
+    '''{docstring}
+    '''
+    return self.{operation_name}(op="{operation}", {arglist})
```

Comment:
I have a couple of questions about this. This generates functions like this:

```python
def min(self, skipna: 'bool' = None, min_count: 'int' = 0, *args, **kwargs):
    '''Compute min of column values.

        skipna : bool
            Whether or not na values must be skipped.
        min_count : int, default 0
            The minimum number of entries for the reduction, otherwise the
            reduction returns NaN.

    '''
    return self._reduce(op="min", skipna=skipna, min_count=min_count, *args, **kwargs)
```
Could it ever get much more complicated than this? I can't put a breakpoint in here anywhere, which is fine as long as this is all that it's doing (just passing along some parameters). Is there any reason we'd have to special case a single function in this area of the code?

As a second question, this approach makes the traceback a little confusing if something goes wrong. But I found out you could use [compile](https://docs.python.org/2/library/functions.html#compile) to name the string just so that it says something a little more informative then `File "<string>", line 12, in sum` if something goes wrong.

Reply:
We discussed this a little in our last team meeting, but to close the loop here, the answer is no this shouldn't get more complicated. If you need to do something more then this mixin is not the appropriate solution. The goal here is to abstract away cases that are exactly this complex, no more. It turns out that this pattern alone is common enough that we don't need to encode any more complexity to be useful, at least not right now.

As to your second point, yes we could compile the string (did you notice that you linked Python2 :P). This point is moot now though because we've moved to a slightly modified approach that uses a descriptor to avoid exec altogether (thanks @shwina for prototyping!) and it's substantially cleaner now.